### PR TITLE
Python 3.7 fixes

### LINKED ISSUE #20690: Invalid syntax error while importing tensorflow in python3.7.0
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
MacOS High Sierra 10.13.5

- **TensorFlow installed from (source or binary)**:
pip3

- **TensorFlow version (use command below)**:
1.9.0

- **Python version**: 
3.7.0

- **Exact command to reproduce**:
import tensorflow as tf

### Describe the problem
when import tensorflow in python3 prompt, it says "SyntaxError: invalid syntax".

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.7/site-packages/tensorflow/__init__.py", line 22, in <module>
    from tensorflow.python import pywrap_tensorflow  # pylint: disable=unused-import
  File "/usr/local/lib/python3.7/site-packages/tensorflow/python/__init__.py", line 49, in <module>
    from tensorflow.python import pywrap_tensorflow
  File "/usr/local/lib/python3.7/site-packages/tensorflow/python/pywrap_tensorflow.py", line 58, in <module>
    from tensorflow.python.pywrap_tensorflow_internal import *
  File "/usr/local/lib/python3.7/site-packages/tensorflow/python/pywrap_tensorflow_internal.py", line 114
    def TFE_ContextOptionsSetAsync(arg1, async):_
                                             ^
SyntaxError: invalid syntax

### Source code / logs
line 114, 115, 150 of pywrap_tensorflow_internal.py has "async" as parameter which seems to be a keyword.
After changed to "async1", importing tensorflow works.

def TFE_ContextOptionsSetAsync(arg1, async1):
    return _pywrap_tensorflow_internal.TFE_ContextOptionsSetAsync(arg1, async1)
TFE_ContextOptionsSetAsync = _pywrap_tensorflow_internal.TFE_ContextOptionsSetAsync

### PR DESCRIPTION
**Pull request by KarlAmort**
### Reserved word: ```async```
"async" became a reserved word in Python 3.7. I replaced with "is_async", which was already used for such flags in other places. An alternative suggestion, if the ```is_``` construct is deemed too ugly, might be to use the full term, "asynchronous".

### Python C API changes

Python C API now returns ```const``` types for `PyUnicode_AsUTF8AndSize``` and similar. Fixed analogous to https://github.com/google/protobuf/commit/0a59054c30e4f0ba10f10acfc1d7f3814c63e1a7.

Fixes #20690

### CODE REVIEW

File: tensorflow/c/eager/c_api.cc [hunk 1]
```diff
@@ -111,7 +111,7 @@ tensorflow::Status GetAllRemoteDevices(
 tensorflow::Status CreateRemoteContexts(
     const std::vector<string>& remote_workers, int64 rendezvous_id,
     const tensorflow::ServerDef& server_def,
-    tensorflow::eager::EagerClientCache* remote_eager_workers, bool async,
+    tensorflow::eager::EagerClientCache* remote_eager_workers, bool is_async,
```
Comment (asimshankar) [inline tensorflow/c/eager/c_api.cc]: If `async` is a reserved word in Python 3.7, I don't quite follow why all these C files have to change :)
(`c_api.cc`, `c_api.h`, `context.cc`, `context.h` etc.)

Let's revert the changes to the C files?

File: tensorflow/python/eager/pywrap_tfe_src.cc [hunk 2]
```diff
@@ -216,7 +216,7 @@ bool ParseStringValue(const string& key, PyObject* py_value, TF_Status* status,
 #if PY_MAJOR_VERSION >= 3
   if (PyUnicode_Check(py_value)) {
     Py_ssize_t size = 0;
-    char* buf = PyUnicode_AsUTF8AndSize(py_value, &size);
+    char* buf = const_cast<char*>(PyUnicode_AsUTF8AndSize(py_value, &size));
```
Comment (asimshankar) [inline tensorflow/python/eager/pywrap_tfe_src.cc]: I'd suggest we avoid the `const_cast` and instead make `buf` a `const char*`:

```c++
const char* buf = PyUnicode_AsUTF8AndSize(py_value, &size);
```

File: tensorflow/python/lib/core/ndarray_tensor.cc [hunk 3]
```diff
@@ -154,7 +154,7 @@ Status PyBytesArrayMap(PyArrayObject* array, F f) {
     if (PyUnicode_Check(item.get())) {
 #if PY_VERSION_HEX >= 0x03030000
       // Accept unicode by converting to UTF-8 bytes.
-      ptr = PyUnicode_AsUTF8AndSize(item.get(), &len);
+      ptr = const_cast<char*>(PyUnicode_AsUTF8AndSize(item.get(), &len));
```
Comment (asimshankar) [inline tensorflow/python/lib/core/ndarray_tensor.cc]: Same comment here. Might have to change the `f` provided at the call sites to use a `const char*` instead, but seems like that should be easy.

File: tensorflow/python/lib/core/py_func.cc [hunk 4]
```diff
@@ -352,7 +352,7 @@ Status ConvertNdarrayToTensor(PyObject* obj, Tensor* ret) {
         Py_ssize_t el_size;
         if (PyBytes_AsStringAndSize(input_data[i], &el, &el_size) == -1) {
 #if PY_MAJOR_VERSION >= 3
-          el = PyUnicode_AsUTF8AndSize(input_data[i], &el_size);
+          el = const_cast<char*>(PyUnicode_AsUTF8AndSize(input_data[i], &el_size));
```
Comment (asimshankar) [inline tensorflow/python/lib/core/py_func.cc]: Avoid `const_cast` here too?

**Review by asimshankar — CHANGES_REQUESTED** [review]
Thanks for the PR. Some questions/comments:

File: tensorflow/contrib/distribute/python/mirrored_strategy.py [hunk 5]
```diff
@@ -193,7 +193,7 @@ def distribute_dataset(self, dataset_fn):
         self._prefetch_on_device)
 
   def _broadcast(self, tensor, destinations):
-    # TODO(josh11b): In eager mode, use one thread per device, or async mode.
```
Comment (alextp) [inline tensorflow/contrib/distribute/python/mirrored_strategy.py]: This change is also unnecessary as it's in a comment

**Comment by alextp** [comment]
This change only changes C++ files and one python comment to not use a python keyword. So nothing in here is necessary, so I'm closing this PR.

**Comment by asimshankar** [comment]
@alextp - We'd still want to address the change in the return type of `PyUnicode_AsUTF8AndSize` in 3.7. So reopening the PR.

@MatthiasWinkelmann : Can you reduce the PR to handle just that change?

**Comment by TheAifam5** [comment]
Any update ? :D

File: tensorflow/c/eager/c_api.cc (hunk 1 above)
Comment (daa) [reply tensorflow/c/eager/c_api.cc]: ``c_api.cc`` and ``c_api.h`` have to change because ``pywrap_tensorflow_internal.py`` is generated from them, however I found that change should be rather minimal and touch only ``TFE_ContextOptionsSetAsync()`` and ``TFE_ContextSetAsyncForThread()``

Comment (asimshankar) [reply tensorflow/c/eager/c_api.cc]: Shouldn't this be handled in `pywrap_tfe.i` instead of changing the C code?

Comment (daa) [reply tensorflow/c/eager/c_api.cc]: You are right, this may be handled in ``pywrap_tfe.i`` but from my small investigation I could solve this issue only with some sort of hack - define macro to redefine variable name, i.e. put
```
#define async is_async
```
into ``pywrap_tfe.i``. ``%rename`` function didn't help because it works only with declarations but not with arguments.

Comment (asimshankar) [reply tensorflow/c/eager/c_api.cc]: Would writing a wrapper function in the SWIG file work?
(Not for the same reason, but we hide `TF_SessionRun` from Python and instead use `TF_SessionRun_wrapper` in https://github.com/tensorflow/tensorflow/blob/b8a9d163d9cbb4b581c044d9c4b1b256c801a9c4/tensorflow/python/client/tf_session.i#L482 )

Comment (daa) [reply tensorflow/c/eager/c_api.cc]: I thought of wrapping and I think it should work and it is more direct way than defining macro. But it will require watching for changes in tensorflow c api - watching for updated signatures of wrapped functions and for appearing ``async`` parameter in functions.

Comment (daa) [reply tensorflow/c/eager/c_api.cc]: You are right, there are only 2 functions and one should watch for new signatures in case of updated C code anyway.

**Comment by asimshankar** [comment]
Closing due to inactivity, and since #21202 seems like it subsumes this.
